### PR TITLE
Fix missing start on MK3/S

### PR DESCRIPTION
**Pull request by leptun**
In some cases, the bootloader would skip sending "start\n" on reset with LANG_MODE 1.
In LANG_MODE 0 the "start" wasn't sent at all.

PFW-1110

### CODE REVIEW

File: Firmware/optiboot_w25x20cl.cpp [hunk 1]
```diff
@@ -99,9 +99,9 @@ struct block_t;
 extern struct block_t *block_buffer;
 
 //! @brief Enter an STK500 compatible Optiboot boot loader waiting for flashing the languages to an external flash memory.
-void optiboot_w25x20cl_enter()
+uint8_t optiboot_w25x20cl_enter()
```
Comment (DRracer) [inline Firmware/optiboot_w25x20cl.cpp:104]: please document the new return values:
`//! @return 1 if ...`

Comment (leptun) [reply Firmware/optiboot_w25x20cl.cpp:104]: updated

**Comment by leptun** [comment]
Pushing another fix to this PR. In the past because of the Optiboot implementation for flashing the language area, we were skipping a few characters on startup. This caused octoprint to give an error when connecting (resend from printer). That wasn't the end of the world since that error was suppressed when connecting for this exact reason.
```
Connecting to: COM8
Changing monitoring state from "Offline" to "Opening serial port"
Connected to: Serial<id=0x54a9690, open=True>(port='COM8', baudrate=115200, bytesize=8, parity='N', stopbits=1, timeout=10.0, xonxoff=False, rtscts=False, dsrdtr=False), starting monitor
Changing monitoring state from "Opening serial port" to "Connecting"
Recv: start
Send: N0 M110 N0*125
Recv: echo: 3.9.0-3421
Recv: echo: Last Updated: May 27 2020 09:37:04 | Author: (none, default config)
Recv: Compiled: May 27 2020
Recv: echo: Free Memory: 1583  PlannerBufferBytes: 1792
Recv: echo:Stored settings retrieved
Recv: adc_init
Recv: CrashDetect DISABLED
Recv: FSensor ENABLED
Recv: echo:SD init fail
Recv: Error:No Line Number with checksum, Last Line: 0
Recv: Resend: 1
Recv: ok
Changing monitoring state from "Connecting" to "Operational"
Send: N0 M110 N0*125
Recv: echo:SD init fail
Recv: ok
Send: N1 M115*39
Recv: FIRMWARE_NAME:Prusa-Firmware 3.9.0 based on Marlin FIRMWARE_URL:https://github.com/prusa3d/Prusa-Firmware PROTOCOL_VERSION:1.0 MACHINE_TYPE:Prusa i3 MKmini EXTRUDER_COUNT:1 UUID:00000000-0000-0000-0000-000000000000
Recv: ok
Send: M21
Recv: echo:SD init fail
Recv: ok
Send: M105
Recv: ok T:29.5 /0.0 B:29.5 /0.0 T0:29.5 /0.0 @:0 B@:0 P:78.1
```
This resend request becomes very problematic once Octoprint is streaming gcode to the printer. The printer would signal Octoprint to resend line 1, but that is not possible anymore and M112 would be sent to the printer.
With this latest fix we are no longer skipping the few bytes at the beginning of the communication and are able to receive a full command as soon as "start\n" is sent to Octoprint. Here is a log with the fix. Notice how there is no resend request:
```
Connecting to: COM4
Changing monitoring state from "Offline" to "Opening serial port"
Connected to: Serial<id=0x78e9250, open=True>(port='COM4', baudrate=115200, bytesize=8, parity='N', stopbits=1, timeout=10.0, xonxoff=False, rtscts=False, dsrdtr=False), starting monitor
Changing monitoring state from "Opening serial port" to "Connecting"
Recv: start
Send: N0 M110 N0*125
Recv: echo: 3.9.0-3421
Recv: echo: Last Updated: May 27 2020 11:41:13 | Author: (none, default config)
Recv: Compiled: May 27 2020
Recv: echo: Free Memory: 1675  PlannerBufferBytes: 1792
Recv: echo:Stored settings retrieved
Recv: adc_init
Recv: CrashDetect DISABLED
Recv: FSensor ENABLED (sensor board revision: 0.3 or older)
Recv: echo:SD init fail
Recv: echo:SD init fail
Recv: ok
Changing monitoring state from "Connecting" to "Operational"
Send: N0 M110 N0*125
Recv: ok
Send: N1 M115*39
Recv: FIRMWARE_NAME:Prusa-Firmware 3.9.0 based on Marlin FIRMWARE_URL:https://github.com/prusa3d/Prusa-Firmware PROTOCOL_VERSION:1.0 MACHINE_TYPE:Prusa i3 MK3S EXTRUDER_COUNT:1 UUID:00000000-0000-0000-0000-000000000000
Recv: ok
Send: M21
```
For the best Octoprint experience, the user must enable the `Wait for start on connect` setting.

File: Firmware/Marlin_main.cpp [hunk 2]
```diff
@@ -1040,15 +1048,13 @@ void setup()
...
-	
-	selectedSerialPort = eeprom_read_byte((uint8_t*)EEPROM_SECOND_SERIAL_ACTIVE);
-	if (selectedSerialPort == 0xFF) selectedSerialPort = 0;
 	if (farm_mode)
 	{
 		no_response = true; //we need confirmation by recieving PRUSA thx
 		important_status = 8;
 		prusa_statistics(8);
 		selectedSerialPort = 1;
+		MYSERIAL.begin(BAUDRATE);
```
Comment (DRracer) [inline Firmware/Marlin_main.cpp:1057]: why is the serial being initialized again? See line 1020. Is it because the selectedSerialPort may have changed?

Comment (leptun) [reply Firmware/Marlin_main.cpp:1057]: Yes. If selectedSerialPort was disabled prior to the farm_mode check, then the serial would be broken on the first farm mode boot

Comment (leptun) [reply Firmware/Marlin_main.cpp:1057]: When the serial was initialized later in the code this wasn't an issue, but now that we initialize it so early we really need this extra re-init. In any case, it just sets the UART registers again. No buffer manipulation happens.

File: Firmware/Marlin_main.cpp [hunk 3]
```diff
@@ -1014,11 +1014,19 @@ void setup()
 	lcd_splash();
     Sound_Init();                                // also guarantee "SET_OUTPUT(BEEPER)"
 
+	selectedSerialPort = eeprom_read_byte((uint8_t *)EEPROM_SECOND_SERIAL_ACTIVE);
+	if (selectedSerialPort == 0xFF) selectedSerialPort = 0;
+	eeprom_update_byte((uint8_t *)EEPROM_SECOND_SERIAL_ACTIVE, selectedSerialPort);
```
Comment (DRracer) [inline Firmware/Marlin_main.cpp:1019]: Do we really need to update the EEPROM? The 0xff value was handled correctly...

**Review by DRracer — APPROVED** [review]

**Pull request merged**